nexus/src/external_api/http_entrypoints.rs

```diff
@@ -838,10 +1027,46 @@ async fn silo_policy_update(
 // Silo-specific user endpoints
 
 /// List users in a silo
+#[endpoint {
+    method = GET,
+    path = "/v1/system/users",
+    tags = ["system"],
+}]
```

Comment:
cc @davepacheco, @ahl, @jmpesp 

This diverges from the initial implementation as discussed here: https://github.com/oxidecomputer/omicron/pull/1784#discussion_r990515243. 

As per the original implementation this would've been `/v1/system/users/all`. The problem (as I see it) with that approach is that it's divergent from our other endpoint definitions. The stated challenge we were trying to work around is that there are potentially other ways that we may want to look up a user other than ID. While I appreciate that challenge, I think there are other ways we can go about solving it. 

My current suggestion (for looking up a user by `external_id`) is simply to add a filter to the list endpoint. 

```
GET /v1/system/users?external_id={id}`
```

This would yield an array with 0 or 1 results. While not strictly as clean from a result perspective, this approach could be added to other endpoints which already implement ID or name lookups as a subroute of the list endpoint without requiring breaking path changes. 

Bulk lookups were also mentioned. I feel like this could also be handled via filtering the list endpoint. One could envision an array query parameter like `/v1/system/users?ids[]=xxx&ids[]=yyy`.

I acknowledge that there may well be a strong counter argument here that having multiple operating modes in a single endpoint makes that endpoints implementation more complex. While acknowledging that, I'd also argue that we're simply talking about different types of filters here. Generally I think we'll _want_ to give more robust filtering capabilities to users and this'll be something we should work towards adding anyway. 

All that said, feed back is definitely welcome.

Reply:
I have the same problem in #2423 with looking up groups by external ID. I agree with Justin that while it's a bit weird, treating it as a list filter is the best way.

Reply:
I think I initially went with separate endpoints out of my general fear of scalably implementing arbitrary filtering.  More specifically, if we choose _not_ to implement arbitrary filtering, and instead implement only a handful of filters that are important (e.g., users in one group, user for an external id, all users, users for a set of ids), will the resulting endpoint restrict users in ways that they'll find surprising and annoying (like: "I can take a set of user ids or a group id, but not both")?  I'm still worried about arbitrary filtering but I don't have a strong feeling either way on which makes a cleaner API.

Tangentially: I think we have generally avoided structured query parameters (i.e., where the backend synthesizes more complicated structures from the key-value pairs), but maybe that's just because we haven't had enough need for them.  The main downside (to me) is that I don't think there's really a standard for it so it winds up being API-specific instead of predictable or conventional.  But we're probably going to need them at some point so we may as well come up with a standard.  In this case that could be `ids[]=a&ids[]=b` (as you suggested) or `id=a&id=b` or `id='%5Ba%2C%20b%5D` (filter field is url-encoded JSON) or even `filter='%22%7B%5C%22id%5C%22%3A%5B%5C%22a%5C%22%2C%5C%22b%5C%22%5D%7D%22` (filter is url-encoded JSON).

I think I wouldn't worry too much about either of these here.